packages/map/src/CanvasPins.ts

```diff
@@ -0,0 +1,351 @@
+CanvasPins.prototype.publish("bottomRightX", null, "number", "Bottom right x-value", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("bottomRightY", null, "number", "Bottom right y-value", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("topLeftX", null, "number", "Top left x-value", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("topLeftY", null, "number", "Top left y-value", null, { tags: ["Basic"], optional: true });
+
+CanvasPins.prototype.publish("pinHeight", 20, "number", "pinHeight", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("pinWidth", 20, "number", "pinWidth", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("pinFontFamily", "Arial", "string", "pinFontFamily", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("pinFontSize", 14, "number", "pinFontSize", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("arrowHeight", 8, "number", "arrowHeight", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("arrowWidth", 8, "number", "arrowWidth", null, { tags: ["Basic"], optional: true });
+
+export class Quadtree {
+    protected _pin_h = 0;
+    protected _pin_w = 0;
+    protected _visit_count: number = 0;
+    protected _tree: any;
+    constructor(extent: any, raw_data: any, pin_h: number, pin_w: number) {
+        this._pin_h = pin_h;
+        this._pin_w = pin_w;
+        this._tree = d3quadtree()
+            .extent(extent)
+            .addAll(raw_data);
+    }
+
+    searchRect(left: number, top: number, right: number, bottom: number): any {
+        const ret: any = [];
+        const pin_h = this._pin_h;
+        const pin_w = this._pin_w;
+        this._tree.visit((node: any, x1: number, y1: number, x2: number, y2: number) => {
+            this._visit_count++;
+            if (!node.length) {
+                let next_exists = false;
+                do {
+                    const d = node.data;
+                    const half_w = pin_w / 2;
+                    const half_h = pin_h / 2;
+                    const d_left = d[0] - half_w;
+                    const d_right = d[0] + half_w;
+                    const d_top = d[1] - half_h;
+                    const d_bottom = d[1] + half_h;
+                    const left_in_range = (d_left >= left) && (d_left <= right);
+                    const right_in_range = (d_right <= right) && (d_right >= left);
+                    const top_in_range = (d_top >= top) && (d_top <= bottom);
+                    const bottom_in_range = (d_bottom <= bottom) && (d_bottom >= top);
+                    const x_in_range = left_in_range || right_in_range || d_left < left && d_right > right;
+                    const y_in_range = top_in_range || bottom_in_range || d_top < top && d_bottom > bottom;
```

Comment:
Move this logic into its own function `overlaps(r1:RECT, r2:RECT)` or similar (where RECT is a type declaration for your rectangle) - you may find it easier to test if two rectangle *don't* overlap:
```
function overlaps(r1, r2): boolean{
  if (r1.right < r2.left || r2.right < r1.left) return false;
  if (r1.bottom < r2.top || r2.bottom < r1.top) return false;
  return true;
}
```
Type of thing.

Reply:
Mind = blown... such a good point!